### ISSUE TITLE
Fine grained ignores

### ISSUE THREAD

**Issue by kangax** [stylelint/stylelint#685]
I'm running into a limitation with stylelint where I need to ignore certain rules in certain files (in my case we need to allow vendor-prefixed properties in some files). Naturally, I can't just use `ignoreFiles` directive in config as that will ignore those files entirely.

So I'm wondering — what's the best way to implement something like that?

I can obviously write my own custom plugin and basically replicate no-vendor-* rule logic except that I would only do it if a file doesn't have certain declaration (e.g. `/* @vendor-prefix-allowed */`). This is how it's currently done, btw. I really wouldn't want to duplicate already existing rule with just 1 minor addition of *ignoring that rule* if a file has certain string in it.

One option is something like: 

```js
'property-no-vendor-prefix': [true, {
  message: '...',
  warn: true,
  exceptWhenFileContains: '@@vendor-prefix-allowed'
}]
```

..but I certainly understand that this might be too specific of a case to add support for it natively.

I wanted to get this idea out there and pick your brains on this :)

Would you be open to such option or something similar? Can you see another (more elegant/simpler) way to achieve this? Has anyone ever requested something like this? Any other thoughts?

**Comment by davidtheclark** [comment]
@kangax: Does a `stylelint-disable` command work for your use-case? https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#turning-rules-on-and-off 

That's pretty much `stylelint-disable` is intended for, I think. 

So if `stylelint-disable` *doesn't* work for you, could you explain why? — because maybe we could improve it.

**Comment by kangax** [comment]
@davidtheclark stylelint-disable is definitely a fantastic option; but I would need to wrap tons of stuff in it in a bunch of files. It's doable, but is there a better way? After some thinking, it became clear that a good solution would be to reuse existing rules in custom plugins. An inhertiance of a sort.

Check out this example to make things clearer:

```js
'use strict';

const stylelint = require('stylelint');
const ruleName = 'no-vendor-prefix';

const propertyNoVendorPrefixRule =
  require('../node_modules/stylelint/dist/rules/property-no-vendor-prefix/index.js').default;

const valueNoVendorPrefixRule =
  require('../node_modules/stylelint/dist/rules/value-no-vendor-prefix/index.js').default;

const selectorNoVendorPrefixRule =
  require('../node_modules/stylelint/dist/rules/selector-no-vendor-prefix/index.js').default;

module.exports = stylelint.createPlugin(ruleName, function(ruleValue) {

  return function(root, result) {

    const validOptions = stylelint.utils.validateOptions(result, ruleName, {
      actual: ruleValue,
    });
    if (!validOptions) { return; }

    let shouldIgnore;
    root.walkComments(comment => {
      if (comment.toString().indexOf('@no-browser-specific-css') > -1) {
        shouldIgnore = true;
      }
    });

    if (shouldIgnore) { return; }

    propertyNoVendorPrefixRule(ruleValue)(root, result);
    valueNoVendorPrefixRule(ruleValue)(root, result);
    selectorNoVendorPrefixRule(ruleValue)(root, result);
  };
});
```

The implementation is horrible of course. I don't feel comfortable reaching all the way into the module's internal dist/rules and reusing the logic. Ideally it would be exposed through stylelint.

```js
const stylelint = require('stylelint');
const selectorNoVendorPrefixRule = stylelint.rules['no-vendor-prefix-rule'];
...
```

This solves my case elegantly and should be useful for anyone looking to reuse existing rules logic (extending, combining, etc.)

What do you think?

**Comment by davidtheclark** [comment]
At first I was a little confused as to why it would be easier to use the `@no-browser-specific-css` comment than a `stylelint-disable` comment ... but I think I get it now: Sounds like your difficulty is that you don't want to disable specific rules for specific files only, but instead you want to *enable* specific rules for specific files. That is, most of your files will *not* get these vendor-prefix checks; you would like to manually specify which ones *should*. Is that understanding correct?

**Comment by davidtheclark** [comment]
Wait, now on re-reading I'm confused again. Can you clarify: why would your `@no-browser-specific-css` comment be *easier* to apply to your codebase than `stylelint-disable` comments?

**Comment by kangax** [comment]
Actually, you're right that it's almost the same. 

I can replace:

```css
/* @no-browser-specific-css */
...
```
with:
```css
/* stylelint-disable value-no-vendor-prefix, property-no-vendor-prefix, selector-no-vendor-prefix */
...
```

I'll just need to replace this in all the files (as well as in a build tool that we use to transform or not transform CSS) so I'd rather teach stylelint/rule about our directive.

However, it's not a big deal and you're right that stylelint-disable already solves this.

I'm still curious about exposing rules / inheritance. I think it makes a lot of sense to add.

**Comment by davidtheclark** [comment]
I am open to the idea of exposing rules to make something like this feasible. @jeddy3, what do you think? It would allow for much more complex conditional enforcement, custom directives like @kangax's, etc.

**Comment by jeddy3** [comment]
Sounds good to me :)

**Comment by davidtheclark** [comment]
Yes
